### PR TITLE
Configure react app with webpack having separate configuration

### PR DESCRIPTION
**Pull request by Saeeda-kh**
React skeleton app using:
- webpack multi env support.
- scss.
- React HMR.
- Babel multi env config.
- eslint.
- prettier.
- public folder root config.

### CODE REVIEW

File: webpack.common.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+const HtmlWebpackPlugin = require("html-webpack-plugin");
+const MiniCssExtractPlugin = require("mini-css-extract-plugin");
+
+const public = path.resolve(__dirname, "public");
+
+module.exports = {
+  entry: "./src/index.tsx",
+  output: {
+    filename: "[name].[contenthash].js",
+    path: path.resolve(__dirname, "build"),
```
Comment (misbahkhalilaz) [inline webpack.common.js:12]: Resolve this path in var on top, and use it here as you've done with the public path.
```
const __builddir = path.resolve(__dirname, "build");
```

File: webpack.common.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+      {
+        test: /\.scss$/,
+        use: [MiniCssExtractPlugin.loader, "css-loader", "sass-loader"],
+      },
+    ],
+  },
+
+  resolve: {
+    extensions: [".*", ".js", ".jsx", ".ts", ".tsx"],
```
Comment (misbahkhalilaz) [inline webpack.common.js:47]: I think once you've added the `.*` extension then you don't need to specify any ext explicitly.

File: webpack.dev.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+const path = require("path");
+
+module.exports = merge(commonConfig, {
+  mode: "development",
+  devtool: "inline-source-map",
+  devServer: {
+    port: 8080,
+    open: true,
+  },
```
Comment (misbahkhalilaz) [inline webpack.dev.js:15]: Good work Genies.

File: webpack.dev.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    new HtmlWebpackPlugin({
+      template: "public/index.html",
+    }),
+    new ErrorOverlayPlugin(), //it provides an overlay in the browser when there are compilation errors.
+    new webpack.HotModuleReplacementPlugin(), //updates our code without reloading page
+  ],
+
+  output: {
+    publicPath: "/",
+  },
```
Comment (misbahkhalilaz) [inline webpack.dev.js:28]: I think it's already added in the common config. Remove it from here.

**Review by misbahkhalilaz — CHANGES_REQUESTED** [review]

File: package.json [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  "main": "./src/index.tsx",
+  "engines": {
+    "node": "16.15.1",
+    "npm": "8.11.0"
+  },
+  "scripts": {
+    "build": "webpack --config webpack.prod.js",
+    "dev": "webpack-dev-server --hot --config webpack.dev.js --port 3000",
+    "start": "npm run build & serve -s build -p 3010",
+    "test": "test"
```
Comment (misbahkhalilaz) [inline package.json:14]: remove the test command for now, as we've not yet configured any automated testing feature.

File: package.json [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    "typescript": "^5.0.4",
+    "webpack": "^5.68.0",
+    "webpack-cli": "^4.9.2",
+    "webpack-dev-server": "^4.7.4",
+    "css-loader": "^6.7.3",
+    "sass-loader": "^13.2.2"
+  },
+  "dependencies": {
+    "react": "^18.0.0-rc.3",
+    "react-dom": "^18.0.0-rc.3"
```
Comment (misbahkhalilaz) [inline package.json:41]: update these to `18.2.0`.

File: src/index.tsx [hunk 7]
```diff
@@ -0,0 +1,4 @@
+import ReactDOM from 'react-dom';
+import App from './App';
+
+ReactDOM.render(<App/>, document.getElementById('root'))
```
Comment (misbahkhalilaz) [inline src/index.tsx:4]: This method is no longer supported in React v18, update this as:
```tsx
const root = ReactDOM.createRoot(document.getElementById("root") as HTMLElement);
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);
```

File: package.json [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  "description": "A new react app without CRA",
+  "main": "./src/index.tsx",
+  "engines": {
+    "node": "16.15.1",
+    "npm": "8.11.0"
+  },
+  "scripts": {
+    "build": "webpack --config webpack.prod.js",
+    "dev": "webpack-dev-server --hot --config webpack.dev.js --port 3000",
+    "start": "npm run build & serve -s build -p 3010",
```
Comment (misbahkhalilaz) [inline package.json:13]: As we've configured the public folder using Webpack, I think the public folder thing would not work with the `serve` package, check this out if the public folder works fine while serving build with the `serve` package.

File: webpack.dev.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  devtool: "inline-source-map",
+  devServer: {
+    open: true,
+  },
+
+  plugins: [
+    ...commonConfig.plugins,
+    new HtmlWebpackPlugin({        //simplifies the creation of html files. if any assets are changed this plugin auto makes changes in html files
+      template: "public/index.html",
+    }),
```
Comment (misbahkhalilaz) [inline webpack.dev.js:19]: Remove this, as it is already available in common config.

File: webpack.prod.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+module.exports = merge(commonConfig, {
+  mode: "production",
+  plugins: [
+    ...commonConfig.plugins,
+    new HtmlWebpackPlugin({
+      //simplifies the creation of html files. if any assets are changed this plugin auto makes changes in html files
+      template: "public/index.html",
+    }),
+   
+  ],
```
Comment (misbahkhalilaz) [inline webpack.prod.js:13]: This complete config is already available in common config file, remove the `plugin` section.

File: webpack.common.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+      {
+        test: /\.scss$/,
+        use: [MiniCssExtractPlugin.loader, "css-loader", "sass-loader"],
+      },
+    ],
+  },
+
+  resolve: {
+    extensions: [ ".js", ".jsx", ".ts", ".tsx"],
```
Comment (misbahkhalilaz) [inline webpack.common.js:46]: I think `.scss` and `.css` extensions also needs to be added as well.

File: package.json (hunk 5 above)
Comment (Saeeda-kh) [reply package.json:14]: removed

File: package.json (hunk 6 above)
Comment (Saeeda-kh) [reply package.json:41]: updated

File: src/index.tsx (hunk 7 above)
Comment (Saeeda-kh) [reply src/index.tsx:4]: changed

File: webpack.dev.js (hunk 9 above)
Comment (Saeeda-kh) [reply webpack.dev.js:19]: removed

File: webpack.prod.js (hunk 10 above)
Comment (Saeeda-kh) [reply webpack.prod.js:13]: removed

File: webpack.common.js (hunk 11 above)
Comment (Saeeda-kh) [reply webpack.common.js:46]: now added the remaining extensions

**Comment by misbahkhalilaz** [comment]
HMR not working, check this out.

**Review by misbahkhalilaz — APPROVED** [review]

**Pull request merged**